feat(build): Persisting decorator metadata

chuckjaz opened the pull request:
This allows determining what the runtime metadata will be for a class without having to loading and running the corresponding .js file.

Part of #7482

chuckjaz committed:
feat(build): Persisting decorator metadata

This allows determing what the runtime metadata will be for a
class without having to loading and running the corresponding
.js file.

alexeagle commented on tools/broccoli/broccoli-typescript.ts:
```diff
@@ -194,11 +196,26 @@ class DiffingTSCompiler implements DiffingBroccoliPlugin {
       fse.mkdirsSync(path.dirname(absoluteFilePath));
       fs.writeFileSync(absoluteFilePath, this.fixSourceMapSources(fileContent), FS_OPTS);
+      if (endsWith(absoluteFilePath, '.d.ts')) {
+        // TODO: Remove this hack when emit passes us the source file.
```
what does 'when' mean? Is there an issue filed to change the callback from program.emit?

chuckjaz replied:
Added a reference to the typescript issued (after adding it).

alexeagle commented on tools/broccoli/broccoli-typescript.ts:
```diff
@@ -213,8 +230,8 @@ class DiffingTSCompiler implements DiffingBroccoliPlugin {
 
-    if (emitResult.emitSkipped) {
-      let allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
+    if (emitSkipped) {
```
does emitSkipped ever get set true?

chuckjaz replied:
Incomplete revert of a refactoring. Fixed.

alexeagle commented on tools/broccoli/broccoli-typescript.ts:
```diff
@@ -213,8 +230,8 @@ class DiffingTSCompiler implements DiffingBroccoliPlugin {
-    if (emitResult.emitSkipped) {
-      let allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
+    if (emitSkipped) {
+      let allDiagnostics = ts.getPreEmitDiagnostics(program).concat(diagnostics);
```
I don't see where diagnostics is set either

chuckjaz replied:
Incomplete revert of a refactoring. Fixed.

alexeagle commented on tools/broccoli/broccoli-typescript.ts:
```diff
@@ -194,11 +196,26 @@ class DiffingTSCompiler implements DiffingBroccoliPlugin {
 
   private doFullBuild() {
```
do we also need to produce metadata for incremental builds?

chuckjaz replied:
My thinking here is that full build is sufficient because, if you are performing an incremental build, you are not using the metadata either. The metadata is only for projects that are referencing angular 2 from an npm install which they only have the .d.ts files. Adding it would be straight forward, however, so let me know what you think.

alexeagle replied:
I don't think that users referencing our published npm are the only readers of metadata.
I imagine while we are developing angular2, we may change some decorators in a way that affects metadata, then re-run a later offline-compilation build step and the changes don't show up because there wasn't a clean build.
Maybe we could wait until it's a problem, but I'm pretty sure it will be.

chuckjaz replied:
Fixed

alexeagle commented on tools/metadata/symbols.ts:
```diff
@@ -0,0 +1,25 @@
+declare var Map: MapConstructor;
+
+var a: Array<number>;
+export class Symbols {
```
is this the ES6 Symbol type? I'm a little confused by what this does, and how it relates to ts.Symbol

chuckjaz replied:
Added a comment to describe what this class is for.

alexeagle commented on tools/metadata/evaluator.ts:
```diff
@@ -0,0 +1,299 @@
+}
+
+function allNodeChildren(node: ts.Node, cb: (node: ts.Node) => boolean) {
+  return !ts.forEachChild(node, node => !cb(node));
```
doesn't this logic mean "any" ?

chuckjaz replied:
The reason I wrapped this is the somewhat unexpected behavior of forEachChild() in that it stops iterating when the callback returns a truthy value. Since most callbacks return `undefined` this is somewhat justified but it makes an `every` style callback a bit awkward to write.

chuckjaz replied:
I renamed the function and added a comment to describe why it is there.

alexeagle commented on tools/metadata/evaluator.ts:
```diff
@@ -0,0 +1,299 @@
+        }
+      }
+    }
+    return undefined;
```
haha, we should turn on --noImplicitReturns someday and then this will be needed :)

alexeagle commented on tools/metadata/extractor.ts:
```diff
@@ -0,0 +1,93 @@
+  return result;
+}
+
+function moduleNameFromBaseName(moduleFileName: string, baseFileName: string) {
```
missing return type

chuckjaz replied:
Fixed

alexeagle commented on tools/metadata/extractor.ts:
```diff
@@ -0,0 +1,93 @@
+   */
+  public getMetadata(sourceFile: ts.SourceFile, typeChecker: ts.TypeChecker): any {
+    const locals = new Symbols();
+    const moduleNameOf: (fileName: string) => string = fileName =>
```
nit: type should be inferable from the assignment

chuckjaz replied:
Fixed

alexeagle commented on tools/metadata/extractor.ts:
```diff
@@ -0,0 +1,93 @@
+      };
+    }
+
+    let metadata;
```
we are still missing --noImplicitAny in tools, I have a PR out for it, le sigh

chuckjaz replied:
Added explicit any.

alexeagle commented:
cool, I have at least read all of the code now. Where are the test cases? There is a lot of conditional logic in here that can be wrong. What's the plan for finding defects when I make a change in here?

chuckjaz commented:
Test will appear in a follow-up PR.

chuckjaz committed:
chore(build): Added tests for metadata extractor

Adds unit test to metadata extractor classes
Fixes issues found while testing

chuckjaz commented:
Actually decided just to fold the tests into this PR

mgechev commented:
I'm very happy to see the great work and progress with the metadata reader! Are you planning to release it as a standalone module?

alexeagle commented:
still LGTM, ready to merge?

chuckjaz commented:
@mgechev This PR is the writer half. I currently plan on having a PR for the reader as a separate module some time this week.

chuckjaz committed:
chore(refactor): Refactored metadata collector

Renamed MetadataExtractor to MetadataCollector
Reorganized to split src from tests

naomiblack commented:
@chuckjaz are we still waiting before merging this one...?

chuckjaz commented:
Yes. I have a refactor and some additional behavior to add before this goes in.

chuckjaz commented:
@mgechev Plans have changed and instead of moving this out of `tools` we currently plan to publish this directory from `tools` as a npm module in a follow up PR. Does that work for you?

I also have a prototype of a resolver that produces the same information that the `Resolver` does at runtime. I am beginning to clean this up and put it in the form of a PR.

tbosch commented on tools/broccoli/broccoli-typescript.ts:
```diff
@@ -240,6 +259,22 @@ class DiffingTSCompiler implements DiffingBroccoliPlugin {
   }
 
   /**
+   * Emit a .metadata.json file to correspond to the .d.ts file if the module contains classes that
```
Could we name the file `.ng2meta.json` file? Would be more clear, especially in G3...

chuckjaz replied:
As discussed offline, the metadata is not Angular specific as it is a symbolic representation all decorators, not just the Angular decorators.

timothysoehnlin commented on tools/metadata/src/evaluator.ts:
```diff
@@ -0,0 +1,412 @@
+              case ts.SyntaxKind.EqualsEqualsEqualsToken:
+                return <any>left === <any>right;
+              case ts.SyntaxKind.ExclamationEqualsEqualsToken:
+                return <any>left === <any>right;
```
Shouldn't that be `return <any> left !== <any> right;`

chuckjaz replied:
Yes! Good catch.

chuckjaz replied:
Fixed. Added more tests as this pointed out that mine were lacking and found an additional issue which I also fixed.

alexeagle commented on tools/metadata/test/collector.spec.ts:
```diff
@@ -0,0 +1,394 @@
+  });
+});
+
+// TODO: Do not use \` in a template literal as it confuses clang-format
```
you can disable clang-format for a range instead
```
int formatted_code;
// clang-format off
    void    unformatted_code  ;
// clang-format on
void formatted_code_again;
```

chuckjaz replied:
Acknowledged.

alexeagle commented:
still LGTM
fix travis failure? maybe squash to one commit if desired?

chuckjaz commented:
Travis failure was a flake.

mgechev commented on tools/metadata/src/collector.ts:
```diff
@@ -0,0 +1,199 @@
+        data.push(metadata);
+        members[name] = data;
+      }
+      for (const member of classDeclaration.members) {
```
What about the metadata associated to property accessors?

chuckjaz replied:
Reported as #7772. Will submit a follow-up PR.

mgechev commented:
@chuckjaz yes, we'll have direct access to the metadata reader API, so it works perfectly.

mary-poppins commented:
Merging PR #7492 on behalf of @kara to branch [presubmit-kara-pr-7492](https://github.com/angular/angular/tree/presubmit-kara-pr-7492).
